fvm/environment/random_generator.go

```diff
@@ -0,0 +1,142 @@
+	uint64,
+	error,
+) {
+	return parseRestrict1Ret(
+		gen.txnState,
+		trace.FVMEnvRandom,
+		gen.impl.UnsafeRandom)
+}
+
+func NewRandomGenerator(
+	tracer tracing.TracerSpan,
+	stateSnapshot protocol.Snapshot,
+	txId flow.Identifier,
+) RandomGenerator {
+	gen := &randomGenerator{
+		tracer:        tracer,
+		stateSnapshot: stateSnapshot,
+		txId:          txId,
+		isPRGCreated:  false, // PRG is not created
+	}
+
+	return gen
+}
+
+func (gen *randomGenerator) createPRG() (random.Rand, error) {
+	// Use the protocol state source of randomness [SoR] for the current block's
+	// execution
+	source, err := gen.stateSnapshot.RandomSource()
+	// expected errors of RandomSource() are:
+	// - storage.ErrNotFound if the QC is unknown.
+	// - state.ErrUnknownSnapshotReference if the snapshot reference block is unknown
+	// at this stage, snapshot reference block should be known and the QC should also be known,
+	// so no error is expected in normal operations
+	if err != nil {
+		return nil, fmt.Errorf("reading random source from state failed: %w", err)
+	}
+
+	// Use the state/protocol PRG derivation from the source of randomness:
+	//  - for the transaction execution case, the PRG used must be a CSPRG
+	//  - use the state/protocol/prg customizer defined for the execution environment
+	//  - use the transaction ID as an extra diversifier of the CSPRG. Although this
+	//    does not add any extra entropy to the output, it allows creating an independent
+	//    PRG for each transaction.
+	csprg, err := prg.New(source, prg.ExecutionEnvironment, gen.txId[:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a CSPRG from source: %w", err)
+	}
+
+	return csprg, nil
+}
+
+// UnsafeRandom returns a random uint64 using the underlying PRG (currently
+// using a crypto-secure one). This function is not thread safe, due to the gen.prg
+// instance currently used. This is fine because a
+// single transaction has a single RandomGenerator and is run in a single
+// thread.
+func (gen *randomGenerator) UnsafeRandom() (uint64, error) {
+	defer gen.tracer.StartExtensiveTracingChildSpan(
+		trace.FVMEnvRandom).End()
+
+	// PRG creation is only done once.
+	if !gen.isPRGCreated {
+		newPRG, err := gen.createPRG()
+		if err != nil {
+			return 0, err
+		}
+		gen.prg = newPRG
+		gen.isPRGCreated = true
+	}
+
+	buf := make([]byte, 8)
+	gen.prg.Read(buf) // Note: prg.Read does not return error
+	return binary.LittleEndian.Uint64(buf), nil
+}
+
+var _ RandomGenerator = (*dummyRandomGenerator)(nil)
+
+// dummyRandomGenerator implements RandomGenerator and is used
+// for the scripts execution environment
+type dummyRandomGenerator struct{}
+
+func NewDummyRandomGenerator() RandomGenerator {
+	return &dummyRandomGenerator{}
+}
+
+// UnsafeRandom() returns an error because executing scripts
+// does not support randomness APIs.
+func (gen *dummyRandomGenerator) UnsafeRandom() (uint64, error) {
+	return 0, errors.NewOperationNotSupportedError("Random")
+}
```

Comment:
I'm not sure about this.

There was previously a discussion for emitting events:
- emitting events only makes sense in transaction
- so we originally threw an error if someone wanted to emit an event in a script
- but this sometimes forced the developers to write 2 functions with the same purpose, one to call within a transaction, with event emitted, and one to call within a script with no event emission.
- to make this more ergonomic we removed the error from event emission and instead just made the event emission in scripts do nothing.

I feel this is relevant here as well. Any function that contains random will be un-callable from a script. One side of the argument is that this is ok, since you cannot have secure deterministic randomness in a script. However calling a contract function with a script is a also great way to poke and prod it to see what it does both for the contract developer and for someone who is just looking at the contract from the outside. 

Maybe just returning 0 would be better....

I'm ok with this for now, but we should definitely discuss this with the community.

Reply:
Good point! I actually asked about this within the team, but I guess you were ooo at that time.
My reasoning was that if we accept a function to run in a script, its result should be equivalent to when we run it as a transaction. Let me know if this requirement does't make sense. 
To replicate the randomness result in scripts, we are missing the transaction hash. 
We could still return a constant (`0`) or use a PRG based on the random beacon only (without a transaction hash). Both options are fine for me, so feel free to pick either.

Reply:
For now this is ok, but lets open up a discussion with the smart contract developers at some point to see if they are ok with it.

Reply:
For future discussions, just adding that the current PR did not disable randomness for script execution. It maintained the original behaviour in scripts but made it more explicit (using a separate implementation of `randomGenerator` that is used for scripts)

Reply:
cc @turbolent